### PR TITLE
[WIP] Bug fix sampled statistics - average calculus in diagnostics package

### PR DESCRIPTION
**Pull request by patapufate**
This PR:

- Change sampled statictics average to moving average window.
- Use large window sizes for historic stats.
- Cap window size to avoid overflow.

### CODE REVIEW

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h [hunk 1]
```diff
@@ -5,31 +5,59 @@
...
+  {
+    if (sample_count_ < windows_size_)
+    {
+      accumulate_ += sample;
+      average_ = accumulate_ / sample_count_;
+      sample_count_++;
+    }
+    else
+    {
+      accumulate_ = accumulate_ + sample - last_value_;
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:43]: @patapufate it doesn't seem like `last_value_` is ever set? It doesn't seem like you can do this without a circular queue either.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h [hunk 2]
```diff
@@ -5,31 +5,59 @@
...
+{
  public:
-  void reset() {
+
+  SampledStatistics (size_t windows_size){
+    windows_size_ = windows_size;
+  }
+  SampledStatistics (){
+    windows_size_ = 100;
+  }
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:22]: @patapufate do we need a default constructor? If it's arbitrary and it has no use for us, I'd suggest we drop it.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h [hunk 3]
```diff
@@ -5,31 +5,59 @@
...
+namespace diagnostic_tools
+{
 template <typename T>
-class SampledStatistics {
+class SampledStatistics
+{
  public:
-  void reset() {
+
+  SampledStatistics (size_t windows_size){
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:17]: @patapufate make this constructor `explicit`.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h [hunk 4]
```diff
@@ -5,31 +5,59 @@
...
+    windows_size_ = windows_size;
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:18]: @patapufate nit: you may use constructor initializers for this.

**Review by hidmic — CHANGES_REQUESTED** [review]
@patapufate IIRC `last_cycle_period_` still gets reset on every diagnostic update. That defeats the purpose of this, doesn't it?

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 5]
```diff
@@ -86,8 +87,8 @@ class PeriodicEventStatus : public PeriodicDiagnosticTask {
 
  private:
   PeriodicEventStatusParams params_;
-  SampledStatistics<double> last_cycle_period_;
-  SampledStatistics<double> historic_period_;
+  SampledStatistics<double> last_cycle_period_{100};
+  SampledStatistics<double> historic_period_{std::numeric_limits<size_t>::max()};
```
[the remark this answers is not in the document]
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:91]: So, if I understood your comment we should:
1) The windows size for:
- last_cycle_period_
- historic_period_
are setted as parameter in each node (ie, Thruster, pose_estimator and health monitor)
2) These parameter are passed in diagnostic_tools::PeriodicMessageStatusParams constructor.
@hidmic is that correct?

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h (hunk 1 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:43]: Implemented with queue.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h (hunk 2 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:22]: Fixed.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h (hunk 3 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:17]: Done.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h (hunk 4 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:18]: Fixed.

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:91]: That makes the most sense to me, yes. We also have to prevent unnecessary window resets.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h [hunk 6]
```diff
@@ -5,34 +5,55 @@
...
  public:
-  void reset() {
+  explicit SampledStatistics(size_t windows_size) { windows_size_ = windows_size; }
+
+  void reset()
+  {
     average_ = T{};
     minimum_ = T{};
     maximum_ = T{};
-    sample_count_ = 0;
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:18]: @patapufate `reset()` should empty the queue.

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 7]
```diff
@@ -76,18 +76,21 @@ class PeriodicEventStatusParams final {
...
 class PeriodicEventStatus : public PeriodicDiagnosticTask {
  public:
-  PeriodicEventStatus(const std::string &name);
+  PeriodicEventStatus(const std::string &name, const size_t &last_cycle_period_avg_window,
+                      const size_t &historic_period_avg_window);
 
-  PeriodicEventStatus(const std::string &name, PeriodicEventStatusParams params);
+  PeriodicEventStatus(const std::string &name, PeriodicEventStatusParams params,
+                      const size_t &last_cycle_period_avg_window,
+                      const size_t &historic_period_avg_window);
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:84]: @patapufate consider putting window sizes into the params structure and providing reasonable defaults. Also, being PODs, I'd suggest you remove `const` and `&` qualifiers.

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 8]
```diff
@@ -76,18 +76,21 @@ class PeriodicEventStatusParams final {
...
   void tick(const ros::Time &stamp) override;
 
   void run(diagnostic_updater::DiagnosticStatusWrapper &stat) override;
 
  private:
   PeriodicEventStatusParams params_;
-  SampledStatistics<double> last_cycle_period_;
-  SampledStatistics<double> historic_period_;
+  SampledStatistics<size_t> last_cycle_period_;
+  SampledStatistics<size_t> historic_period_;
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:93]: @patapufate hmm, samples are floating point numbers, not integers.

File: health_monitor/src/health_monitor.cpp [hunk 9]
```diff
@@ -65,18 +65,26 @@ HealthMonitor::HealthMonitor(ros::NodeHandle &nodeHandle) : nodeHandle(nodeHandl
...
 
+    diagnostic_tools::PeriodicMessageStatusParams paramsReportsHealthMonitorCheckPeriod{};
+    paramsReportsHealthMonitorCheckPeriod.min_acceptable_period(minReportFaultRate);
+    paramsReportsHealthMonitorCheckPeriod.max_acceptable_period(maxReportFaultRate);
     diagnosticsUpdater.add(publisher_reportFault.add_check<diagnostic_tools::PeriodicMessageStatus>(
-        "rate check", diagnostic_tools::PeriodicMessageStatusParams{}
-                          .min_acceptable_period(minReportFaultRate)
-                          .max_acceptable_period(maxReportFaultRate)));
+        "rate check", paramsReportsHealthMonitorCheckPeriod, lastCyclePeriodAverageWindow,
+        historicPeriodAverageWindow));
```
Comment (hidmic) [inline health_monitor/src/health_monitor.cpp:87]: @patapufate why this change? I'd suggest we don't expose them as ROS parameters until we have the need to do so.

**Review by hidmic — CHANGES_REQUESTED** [review]

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 10]
```diff
@@ -66,9 +66,25 @@ class PeriodicEventStatusParams final {
 
   const Diagnostic &stale_diagnostic() const { return stale_diagnostic_; }
 
+  PeriodicEventStatusParams &last_cycle_period_avg_window(double last_cycle_period_avg_window) {
+    last_cycle_period_avg_window_ = last_cycle_period_avg_window;
+    return *this;
+  }
+
+  double last_cycle_period_avg_window() const {return last_cycle_period_avg_window_;}
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:74]: @patapufate Hmm, in what units are window sizes in?

Comment (hidmic) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:74]: @patapufate also, nit: 
```suggestion
  double last_cycle_period_avg_window() const { return last_cycle_period_avg_window_; }
```

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 11]
```diff
@@ -66,9 +66,25 @@ class PeriodicEventStatusParams final {
...
+
+    PeriodicEventStatusParams &historic_period_avg_window(double historic_period_avg_window) {
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:76]: @patapufate nit:
```suggestion
  PeriodicEventStatusParams &historic_period_avg_window(double historic_period_avg_window) {
```

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 12]
```diff
@@ -66,9 +66,25 @@ class PeriodicEventStatusParams final {
...
+    historic_period_avg_window_ = historic_period_avg_window;
+    return *this;
+  }
+  
+  double historic_period_avg_window() const {return historic_period_avg_window_;}
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:81]: @patapufate nit: 
```suggestion
  double historic_period_avg_window() const { return historic_period_avg_window_; }
```

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h [hunk 13]
```diff
@@ -5,34 +5,62 @@
...
-  void reset() {
+  explicit SampledStatistics(T windows_size) { windows_size_ = windows_size; }
+
+  void reset()
+  {
     average_ = T{};
     minimum_ = T{};
     maximum_ = T{};
-    sample_count_ = 0;
+    clear_buffer(buffer_);
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:25]: @patapufate why make a public method for this instead of inlining it?

File: diagnostics_tools/src/periodic_event_status.cpp [hunk 14]
```diff
@@ -4,12 +4,14 @@
...
 PeriodicEventStatus::PeriodicEventStatus(const std::string& name)
-    : PeriodicEventStatus(name, PeriodicEventStatusParams{}) {}
+    : PeriodicEventStatus(name, PeriodicEventStatusParams{}){}
 
 PeriodicEventStatus::PeriodicEventStatus(const std::string& name, PeriodicEventStatusParams params)
-    : PeriodicDiagnosticTask(name), params_(std::move(params)) {}
+    : PeriodicDiagnosticTask(name),
+      params_(std::move(params)),
+      last_cycle_period_(params.last_cycle_period_avg_window()),
+      historic_period_(params.historic_period_avg_window()){}
```
Comment (hidmic) [inline diagnostics_tools/src/periodic_event_status.cpp:14]: @patapufate as far as I can see, this diagnostic check still resets windows on every tick.

File: health_monitor/src/health_monitor.cpp [hunk 15]
```diff
@@ -65,18 +65,25 @@ HealthMonitor::HealthMonitor(ros::NodeHandle &nodeHandle) : nodeHandle(nodeHandl
...
     double maxReportFaultRate = minReportFaultRate * 2.0;
+    int lastCyclePeriodAverageWindow = 100;
+    int historicPeriodAverageWindow = 1000;
     nodeHandle.getParam("/health_monitor_node/report_faults_rate", reportFaultsRate);
     nodeHandle.getParam("/health_monitor_node/min_report_faults_rate", minReportFaultRate);
     nodeHandle.getParam("/health_monitor_node/max_report_faults_rate", maxReportFaultRate);
+    nodeHandle.getParam("/health_monitor_node/last_cycle_period_avg_window",
+                        lastCyclePeriodAverageWindow);
+    nodeHandle.getParam("/health_monitor_node/historic_period_avg_window",
+                        historicPeriodAverageWindow);
```
Comment (hidmic) [inline health_monitor/src/health_monitor.cpp:76]: @patapufate unused parameters?

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 16]
```diff
@@ -66,9 +66,25 @@ class PeriodicEventStatusParams final {
...
+
  private:
   double min_acceptable_period_{-1};
   double max_acceptable_period_{5};
+  double last_cycle_period_avg_window_{1};
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:86]: @patapufate nit: it seems to me that `last_cycle` and `historic` are no longer adequate names. Perhaps `short_term` and `long_term`, respectively? I'm open to other alternatives.

**Review by hidmic — CHANGES_REQUESTED** [review]

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h (hunk 11 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:76]: Fixed

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h (hunk 12 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:81]: Fixed

File: diagnostics_tools/src/periodic_event_status.cpp (hunk 14 above)
Comment (patapufate) [reply diagnostics_tools/src/periodic_event_status.cpp:14]: Only reset if "Time went backwards"

File: health_monitor/src/health_monitor.cpp (hunk 15 above)
Comment (patapufate) [reply health_monitor/src/health_monitor.cpp:76]: Yes.

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h (hunk 16 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:86]: Changed.

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h (hunk 10 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:74]: The Window sizes units are in double.

File: diagnostics_tools/src/periodic_event_status.cpp (hunk 14 above)
Comment (patapufate) [reply diagnostics_tools/src/periodic_event_status.cpp:14]: Although I removed from run method.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h (hunk 13 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:25]: You're right.

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 17]
```diff
@@ -66,9 +66,26 @@ class PeriodicEventStatusParams final {
...
 
+  PeriodicEventStatusParams &short_term_avg_window(double short_term_avg_window) {
+    short_term_avg_window_ = short_term_avg_window;
+    return *this;
+  }
+
+  double short_term_avg_window() const { return short_term_avg_window_; }
+
+  PeriodicEventStatusParams &long_term_avg_window(double long_term_avg_window)
+  {
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:77]: @patapufate nit^2: be consistent with the style (curly brace on the next line vs curly brace on the same line).

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h (hunk 13 above)
Comment (hidmic) [reply diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:25]: No change?

File: diagnostics_tools/src/periodic_event_status.cpp [hunk 18]
```diff
@@ -4,34 +4,36 @@
...
+      long_term_period_(params.long_term_avg_window()){}
+long_term_avg_window
 void PeriodicEventStatus::tick(const ros::Time& stamp) {
   std::lock_guard<std::mutex> guard(mutex_);
   if (!last_stamp_.isZero()) {
     if (stamp < last_stamp_) {
       ROS_WARN_NAMED("diagnostics_tools", "Time went backwards from %f to %f, resetting.",
                      last_stamp_.toSec(), stamp.toSec());
-      last_cycle_period_.reset();
+      short_term_period_.reset();
```
Comment (hidmic) [inline diagnostics_tools/src/periodic_event_status.cpp:22]: @patapufate still resetting?

File: health_monitor/src/health_monitor.cpp [hunk 19]
```diff
@@ -65,6 +65,8 @@ HealthMonitor::HealthMonitor(ros::NodeHandle &nodeHandle) : nodeHandle(nodeHandl
     double reportFaultsRate = 1.0;
     double minReportFaultRate = minReportFaultRate / 2.0;
     double maxReportFaultRate = minReportFaultRate * 2.0;
+    int lastCyclePeriodAverageWindow = 100;
+    int historicPeriodAverageWindow = 1000;
```
Comment (hidmic) [inline health_monitor/src/health_monitor.cpp:69]: @patapufate still necessary?

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 20]
```diff
@@ -66,9 +66,26 @@ class PeriodicEventStatusParams final {
...
+    return *this;
+  }
+
+  double long_term_avg_window() const { return long_term_avg_window_; }
+
  private:
   double min_acceptable_period_{-1};
   double max_acceptable_period_{5};
+  double short_term_avg_window_{1};
+  double long_term_avg_window_{10000};
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:88]: @patapufate these members should be of `size_t` type.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h (hunk 13 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:25]: I declared the function inline. Now I moved the function into the reset method.

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h (hunk 17 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:77]: Ok.

File: health_monitor/src/health_monitor.cpp (hunk 19 above)
Comment (patapufate) [reply health_monitor/src/health_monitor.cpp:69]: Fixed

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h (hunk 20 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:88]: Fixed.

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h [hunk 21]
```diff
@@ -5,46 +5,71 @@
...
   T minimum() const { return minimum_; }
 
- private:
+private:
   T average_{};
   T minimum_{};
   T maximum_{};
-  size_t sample_count_{0};
+  T accumulate_{0};
+  std::queue<T> buffer_;
```
Comment (hidmic) [inline diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:71]: @patapufate FYI there's also [`std::deque`](https://en.cppreference.com/w/cpp/container/deque) which allows in-place clearing.

File: health_monitor/src/health_monitor.cpp [hunk 22]
```diff
@@ -73,10 +73,11 @@ HealthMonitor::HealthMonitor(ros::NodeHandle &nodeHandle) : nodeHandle(nodeHandl
...
                                                &HealthMonitor::reportFaultsTimeout, this);
 
+    diagnostic_tools::PeriodicMessageStatusParams paramsReportsHealthMonitorCheckPeriod{};
+    paramsReportsHealthMonitorCheckPeriod.min_acceptable_period(minReportFaultRate);
+    paramsReportsHealthMonitorCheckPeriod.max_acceptable_period(maxReportFaultRate);
     diagnosticsUpdater.add(publisher_reportFault.add_check<diagnostic_tools::PeriodicMessageStatus>(
-        "rate check", diagnostic_tools::PeriodicMessageStatusParams{}
-                          .min_acceptable_period(minReportFaultRate)
-                          .max_acceptable_period(maxReportFaultRate)));
+        "rate check", paramsReportsHealthMonitorCheckPeriod));
```
Comment (hidmic) [inline health_monitor/src/health_monitor.cpp:80]: @patapufate nit: this is unrelated, but innocuous.

Comment (patapufate) [reply health_monitor/src/health_monitor.cpp:80]: I put it in that way because roslint

File: diagnostics_tools/include/diagnostic_tools/sampled_statistics.h (hunk 21 above)
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/sampled_statistics.h:71]: Good one.

File: diagnostics_tools/include/diagnostic_tools/periodic_event_status.h [hunk 23]
```diff
@@ -66,9 +66,25 @@ class PeriodicEventStatusParams final {
...
+  }
+
+  double short_term_avg_window() const { return short_term_avg_window_; }
+
+  PeriodicEventStatusParams &long_term_avg_window(double long_term_avg_window) {
+    long_term_avg_window_ = long_term_avg_window;
+    return *this;
+  }
+
+  double long_term_avg_window() const { return long_term_avg_window_; }
```
[the remark this answers is not in the document]
Comment (patapufate) [reply diagnostics_tools/include/diagnostic_tools/periodic_event_status.h:81]: Done

**Review by hidmic — APPROVED** [review]
LGTM !
